File: src/BehaviorsSDKNative/Microsoft.Xaml.Interactivity/ActionCollection.cpp
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
#include "pch.h"
#include "ActionCollection.h"

namespace winrt
{
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml;
} // namespace winrt

namespace winrt::Microsoft::Xaml::Interactivity::implementation
{
ActionCollection::ActionCollection()
{
    _vectorChanged = VectorChanged(auto_revoke, {get_weak(), &ActionCollection::OnVectorChanged});
}

void ActionCollection::OnVectorChanged(IObservableVector<DependencyObject> const &, IVectorChangedEventArgs const &event)
{
    auto collectionChange = event.CollectionChange();

    if (collectionChange == CollectionChange::Reset)
    {
        for (auto const &item : *this)
        {
            ActionCollection::VerifyType(item);
        }
    }
    else if (collectionChange == CollectionChange::ItemInserted || collectionChange == CollectionChange::ItemChanged)
    {
        auto changedItem = GetAt(event.Index());
        ActionCollection::VerifyType(changedItem);
    }
}

void ActionCollection::VerifyType(DependencyObject const &item)
{
    auto type = item.try_as<IAction>();
    if (type == nullptr)
    {
        // ResourceHelper::GetString("NonActionAddedToActionCollectionExceptionMessage")
        throw winrt::hresult_invalid_argument(L"NonActionAddedToActionCollectionExceptionMessage");
    }
}
} // namespace winrt::Microsoft::Xaml::Interactivity::implementation

File: src/BehaviorsSDKManaged/Microsoft.Xaml.Interactivity.Shared/ActionCollection.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using Windows.Foundation.Collections;

#if WinUI
using Microsoft.UI.Xaml;
#else
using Windows.UI.Xaml;
#endif

namespace Microsoft.Xaml.Interactivity;

/// <summary>
/// Represents a collection of IActions.
/// </summary>
public sealed partial class ActionCollection : DependencyObjectCollection
{
    /// <summary>
    /// Initializes a new instance of the <see cref="ActionCollection"/> class.
    /// </summary>
    public ActionCollection()
    {
        this.VectorChanged += this.ActionCollection_VectorChanged;
    }

    private void ActionCollection_VectorChanged(IObservableVector<DependencyObject> sender, IVectorChangedEventArgs eventArgs)
    {
        CollectionChange collectionChange = eventArgs.CollectionChange;

        if (collectionChange == CollectionChange.Reset)
        {
            foreach (DependencyObject item in this)
            {
                ActionCollection.VerifyType(item);
            }
        }
        else if (collectionChange == CollectionChange.ItemInserted || collectionChange == CollectionChange.ItemChanged)
        {
            DependencyObject changedItem = this[(int)eventArgs.Index];
            ActionCollection.VerifyType(changedItem);
        }
    }

    private static void VerifyType(DependencyObject item)
    {
        if (!(item is IAction))
        {
            throw new InvalidOperationException(ResourceHelper.GetString("NonActionAddedToActionCollectionExceptionMessage"));
        }
    }
}
